packages/market-service/src/coingecko/coingecko.ts

```diff
@@ -38,6 +43,41 @@ const coingeckoIDMap: CoinGeckoIDMap = Object.freeze({
 export class CoinGeckoMarketService implements MarketService {
   baseUrl = 'https://api.coingecko.com/api/v3'
 
+  private readonly defaultGetByMarketCapArgs: GetByMarketCapArgs = {
+    pages: 10,
+    perPage: 250
+  }
+
+  async getByMarketCap(args?: GetByMarketCapArgs) {
+    const argsToUse = args ?? this.defaultGetByMarketCapArgs
+    const { pages, perPage } = argsToUse
+    const urlAtPage = (page: number) =>
+      `${this.baseUrl}/coins/markets?vs_currency=usd&order=market_cap_desc&per_page=${perPage}&page=${page}&sparkline=false`
+    const pageCount = Array(pages)
+      .fill(0)
+      .map((_v, i) => i + 1)
+    const combined = (
+      await Promise.all(
+        pageCount.map(async (page) => axios.get<CoinGeckoMarketCap>(urlAtPage(page)))
+      )
+    ).flat()
+    return combined
+      .map(({ data }) => data ?? []) // filter out rate limited results
+      .flat()
+      .sort((a, b) => (a.market_cap_rank > b.market_cap_rank ? 1 : -1))
+      .reduce((acc, cur) => {
```

Comment:
except the sort, this could all be done in a single reduce.

Reply:
agreed, i think this is a tad more reasonable. map is only operating on 10 items, flat is native, sort logic is very simple, and it keeps the biz logic to the reduce.